node/src/components/contract_runtime.rs

```diff
@@ -548,65 +567,67 @@ impl ContractRuntime {
     }
 
     pub(crate) fn set_initial_state(&mut self, sequential_block_state: ExecutionPreState) {
-        self.execution_pre_state = sequential_block_state;
+        *self.execution_pre_state.lock().unwrap() = sequential_block_state;
     }
 
-    fn execute_finalized_block<REv>(
-        &mut self,
+    async fn execute_finalized_block_or_requeue<REv>(
+        engine_state: Arc<EngineState<LmdbGlobalState>>,
+        metrics: Arc<ContractRuntimeMetrics>,
+        exec_queue: ExecQueue,
+        execution_pre_state: Arc<Mutex<ExecutionPreState>>,
         effect_builder: EffectBuilder<REv>,
         protocol_version: ProtocolVersion,
-        execution_pre_state: ExecutionPreState,
         finalized_block: FinalizedBlock,
         deploys: Vec<Deploy>,
-    ) -> Effects<ContractRuntimeRequest>
-    where
+    ) where
         REv: From<ContractRuntimeRequest>
             + From<ContractRuntimeAnnouncement>
             + From<ControlAnnouncement>
             + Send,
     {
+        let current_execution_pre_state = { execution_pre_state.lock().unwrap().clone() };
         let BlockAndExecutionEffects {
             block,
             execution_results,
             maybe_step_execution_effect,
-        } = match operations::execute_finalized_block(
-            self.engine_state.as_ref(),
-            self.metrics.as_ref(),
+        } = match tokio::task::unconstrained(operations::execute_finalized_block(
+            engine_state.as_ref(),
+            metrics.as_ref(),
             protocol_version,
-            execution_pre_state,
+            current_execution_pre_state,
             finalized_block,
             deploys,
-        ) {
+        ))
+        .await
```

Comment:
To elaborate a little on @goral09 's point [below][1], we could have this be `async move { operations::execute_finalized_block(...) }` right?

[1]: https://github.com/casper-network/casper-node/pull/1908/files#r688164423